src/aspire/utils/blk_diag_matrix.py

```diff
@@ -0,0 +1,700 @@
+        """
+        Convert full from list representation into BlkDiagMatrix
+
+        :param blk_diag; The blk_diag representation in the form of a
+        K-element list storing all shapes of K diagonal matrix blocks,
+        where `blk_partition[i]` corresponds to the shape (number of rows
+        and columns) of the `i` diagonal matrix block.
+
+        :return: The BlkDiagMatrix matrix
+        """
+
+        # TODO: maybe can improve implementation
+
+        # get the partition (just sizes)
+        partition = BlkDiagMatrix.get_partition(blk_diag)
+
+        # instantiate an empty BlkDiagMatrix with that structure
+        A = BlkDiagMatrix(len(partition), partition=partition, dtype=dtype)
+
+        # set the data
+        for i in range(A.nblocks):
+            A.data[i] = blk_diag[i].copy()
+
+        return A
+
+    @staticmethod
+    def from_mat(mat, blk_partition, dtype=np.float64):
+        """
+        Convert full block diagonal matrix into list representation
+
+        :param mat; The full block diagonal matrix including the zero elements
+        ofnon-diagonal blocks.
+        :param blk_partition: The matrix block partition in the form of a
+        K-element list storing all shapes of K diagonal matrix blocks,
+        where `blk_partition[i]` corresponds to the shape (number rows
+        and columns) of the `i` diagonal matrix block.
+
+        :return: The BlkDiagMatrix matrix
+        """
+
+        # TODO: maybe can improve this implementation
+
+        A = BlkDiagMatrix(len(blk_partition), dtype=dtype)
+
+        rows = blk_partition[:, 0]
+        cols = blk_partition[:, 1]
+        cellarray = Cell2D(rows, cols, dtype=mat.dtype)
+        blk_diag = cellarray.mat2blk_diag(mat, rows, cols)
+        A.data = BlkDiagMatrix.from_blk_diag(blk_diag)
+        return A
+
+    def solve(self, Y):
+        """
+        Solve a linear system involving a block diagonal matrix
+
+        :param Y: The right-hand side in the linear system.  May be a matrix
+        consisting of  coefficient vectors, in which case each column is
+        solved for separately.
+
+        :return: The result of solving the linear system formed by the matrix.
+        """
+
+        rows = np.array([np.size(mat_a, 0) for mat_a in self])
+        if sum(rows) != np.size(Y, 0):
+            raise RuntimeError('Sizes of `self` and `Y` are not compatible.')
+
+        vector = False
+        if np.ndim(Y) == 1:
+            Y = Y[:, np.newaxis]
+            vector = True
+
+        cols = np.array([np.size(Y, 1)])
+        cellarray = Cell2D(rows, cols, dtype=Y.dtype)
+        Y = cellarray.mat2cell(Y, rows, cols)
+        X = []
+        for i in range(0, self.nblocks):
+            X.append(solve(self[i], Y[i]))
+        X = np.concatenate(X, axis=0)
+
+        if vector:
+            X = X[:, 0]
+
+        return X
+
+    def apply(self, X):
+        """
+        Define the apply option of a block diagonal matrix with a matrix of
```

Comment:
Fair point. We can.  Probably it would be natural to group that together with supporting non square blocks. I'll leave this open until I make an issue and tag it. Thanks.

Reply:
Yes let's make an issue.

Reply:
tagging #155